### PR TITLE
Rewrite everything to include epochs logic.

### PR DESCRIPTION
**Pull request by kwix**

### CODE REVIEW

**Comment by kwix** [comment]
Disclaimer: There's more work to be done, especially on tests.

File: contracts/Staking.sol [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        uint min = 0;
+        uint max = checkpoints.length - 1;
+
+        // shortcut for blocks newer than the latest checkpoint == current balance
+        if (epochId >= checkpoints[max].epochId) {
+            return checkpoints[max].balance;
+        }
+
+        // binary search of the value in the array
+        while (max > min) {
```
Comment (miladmostavi) [inline contracts/Staking.sol:138]: This still makes me anxious, but given that it's per user and won't affect everyone, and the number of checkpoints for a user is limited, we should be ok.

**Review by miladmostavi — APPROVED** [review]

File: contracts/Staking.sol [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    function deposit(address tokenAddress, uint256 amount) public {
+        require(amount > 0, "Staking: Amount must be > 0");
+
+        IERC20 token = IERC20(tokenAddress);
+        uint256 allowance = token.allowance(msg.sender, address(this));
+        require(allowance >= amount, "Staking: Token allowance too small");
+
+        balances[msg.sender][tokenAddress] += amount;
+
+        token.transferFrom(msg.sender, address(this), amount);
```
Comment (abancu) [inline contracts/Staking.sol:53]: I believe transferFrom should be above balances+= amount. Also please use safemath.

Comment (abancu) [reply contracts/Staking.sol:53]: It was a bug some time ago related to this.

File: contracts/Staking.sol [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+     * Stores `amount` of `tokenAddress` tokens for the `user` into the vault
+     */
+    function deposit(address tokenAddress, uint256 amount) public {
+        require(amount > 0, "Staking: Amount must be > 0");
+
+        IERC20 token = IERC20(tokenAddress);
+        uint256 allowance = token.allowance(msg.sender, address(this));
+        require(allowance >= amount, "Staking: Token allowance too small");
+
+        balances[msg.sender][tokenAddress] += amount;
```
Comment (abancu) [inline contracts/Staking.sol:51]: please use safemath

File: contracts/Staking.sol [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        uint256 nextEpoch = currentEpoch + 1;
+
+        // update the next epoch pool size
+        Pool storage pNextEpoch = poolSize[tokenAddress][nextEpoch];
+        pNextEpoch.size = token.balanceOf(address(this));
+        pNextEpoch.set = true;
+
+        Checkpoint[] storage checkpoints = balanceCheckpoints[msg.sender][tokenAddress];
+        if ((checkpoints.length == 0) || (checkpoints[checkpoints.length - 1].epochId != nextEpoch)){
+            checkpoints.push(Checkpoint(nextEpoch, balances[msg.sender][tokenAddress]));
```
Comment (abancu) [inline contracts/Staking.sol:66]: again. before pusjing the nextepoch checkpoint we need to check that checkpoints[checkpoints.length - 1].epochId != currentEpoch because otherwise if you get a withdraw just a moment after a deposit the note written in the withdraw is not valid anymore (the user has a balance but for the next epoch not for the current)

Comment (kwix) [reply contracts/Staking.sol:73]: Good catch! Fixed this in the latest version, added a test case for it and also fixed a related edge case.

File: contracts/Staking.sol (hunk 1 above)
Comment (kwix) [reply contracts/Staking.sol:159]: Yeah, it's per user **and per epoch** so we limit it quite a lot. It's going to be a long time before it ends up using a lot of gas.

File: contracts/Staking.sol (hunk 2 above)
Comment (kwix) [reply contracts/Staking.sol:53]: Usually that's how I would have done it but in this case I was aiming for the checks-effects-interactions pattern which avoids re-entrancy attacks. I can change it if there was a different kind of bug affecting it.

File: contracts/Staking.sol (hunk 3 above)
Comment (kwix) [reply contracts/Staking.sol]: done

File: contracts/Staking.sol (hunk 2 above)
Comment (abancu) [reply contracts/Staking.sol:55]: just to be sure we could use OpenZeppelin ReentrancyGuard and add the modifier to this method, this way you could follow the current way to do it.

Comment (kwix) [reply contracts/Staking.sol:54]: done!
